File: rust/src/topcoder/crazybot/crazybot.rs
use std;

pub fn answer() {

    struct direction {
        x: i32,
        y: i32,
        prob: f32
    }

    struct crazy_bot {
        grid: [[i32; 100]; 100],
        dir: [direction; 4]
    }

    impl crazy_bot {
        fn update(&mut self, x: usize, y: usize, num: i32) {
            self.grid[x as usize][y as usize] = num;
        }
    }

    fn makeCb(east: i32, weast: i32, south: i32, north: i32) -> crazy_bot {
        let dir1 = direction{x: 1, y:0, prob: east as f32/100.0};
        let dir2 = direction{x: -1, y:0, prob: weast as f32/100.0};
        let dir3 = direction{x: 0, y:1, prob: south as f32/100.0};
        let dir4 = direction{x: 0, y:-1, prob: north as f32/100.0};

        let mut grid = [[0; 100]; 100];
        let mut dir = [dir1, dir2, dir3, dir4];
        let mut ret = crazy_bot{grid: grid, dir: dir};
        return ret
    }

    fn solve(cb: &mut crazy_bot, n: i32, x:i32, y:i32) -> f32 {
        if(cb.grid[x as usize][y as usize] == 1){
            0.0
        } else if(n == 0) {
            1.0
        } else {

            let mut ret: f32 = 0.0;
            cb.update(x as usize, y as usize, 1);
            let mut i = 0;
            while(i < 4){
                let k = x+cb.dir[i].x;
                let l = y+cb.dir[i].y;
                ret += solve(cb, n-1, k, l) * cb.dir[i].prob;
                i = i + 1
            }
            cb.update(x as usize, y as usize, 0);
            ret
        }

    }

    fn read<T: std::str::FromStr>() -> T {
        let mut s = String::new();
        std::io::stdin().read_line(&mut s).ok();
        s.trim().parse().ok().unwrap()
    }

    fn main() {
        let n = read::<i32>();
        let east = read::<i32>();
        let west = read::<i32>();
        let south = read::<i32>();
        let north = read::<i32>();

        let mut cb = &mut makeCb(east, west, south, north);
        println!("{}", solve(cb, n, 50, 50))
    }
    main()
}

File: haskell/src/TopCoder/CrazyBot/CrazyBot.hs
module TopCoder.CrazyBot.CrazyBot where
import Data.List as L

type Position = (Int, Int)
type Direction = (Position, Float)

dirs :: Int -> Int -> Float -> Direction
dirs x y p = ((x, y), p/100)

solve :: [Position] -> [Direction] -> Int -> Position -> Float
solve hist dirs n p = do
  case p `elem` hist of
    True -> 0.0
    False ->
      if n == 0 then 1
      else L.foldl' (\acc ((nx, ny), prob) -> acc + (solve (hist ++ [((fst p), (snd p))]) dirs (n-1) ((fst p)+nx, (snd p)+ny)) * prob) 0 dirs

answer = do
  n<-L.map read . words<$>getLine::IO[Int]
  east<-L.map read . words<$>getLine::IO[Float]
  west<-L.map read . words<$>getLine::IO[Float]
  south<-L.map read . words<$>getLine::IO[Float]
  north<-L.map read . words<$>getLine::IO[Float]
  let dir = [dirs 1 0 (head east), dirs (-1) 0 (head west), dirs 0 1 (head south), dirs 0 (-1) (head north)]
  print $ solve [] dir (head n) (50, 50)
